src/fsm.jl

```diff
@@ -0,0 +1,229 @@
+    push!(fields, :(prevstate::Fjage.FSMState = Fjage.INIT))
+    push!(fields, :(nextstate::Fjage.FSMState = Fjage.INIT))
+    push!(fields, :(wakers::Vector{Fjage.WakerBehavior} = Fjage.WakerBehavior[]))
+    :( Base.@kwdef mutable struct $T <: Fjage.FSMBehavior; $(fields...); end ) |> esc
+  else
+    @error "Bad FSM definition"
+  end
+end
+
+# standard behavior methods
+done(b::FSMBehavior) = b.state == FINAL
+stop(b::FSMBehavior) = nextstate!(b, FINAL)
+
+function reset(b::FSMBehavior)
+  for waker ∈ b.wakers
+    stop(waker)
+  end
+  empty!(b.wakers)
+  b.agent === nothing || delete!(b.agent._behaviors, b)
+  b.timer === nothing || close(b.timer)
+  b.agent = nothing
+  b.block = nothing
+  b.timer = nothing
+  b.state = INIT
+  b.prevstate = INIT
+  b.nextstate = INIT
+  nothing
+end
+
+"""
+    state(b::FSMBehavior)
+
+Get the current state of the FSM.
+"""
+state(b::FSMBehavior) = b.state
+
+"""
+    reenterstate(b::FSMBehavior)
+
+Reenter the current state of the FSM.
+"""
+reenterstate(b::FSMBehavior) = nextstate!(b, REENTER)
+
+"""
+    nextstate!(b::FSMBehavior, state::FSMState)
+
+Set the next state of the FSM. Transition occurs after the current action
+is completed, if called from the action method for a state.
+"""
+function nextstate!(b::FSMBehavior, state::FSMState)
+  b.nextstate = state
+  restart(b)
+end
+
+"""
+    trigger!(b::FSMBehavior, event)
+
+Trigger an event in the FSM.
+"""
+trigger!(b::FSMBehavior, event) = _mutex_call(onevent, b.agent, b, b.state, event)
+
+"""
+    after(action, b::FSMBehavior, delay)
+
+Schedule an action to be executed after a delay (in seconds). The scheduled action
+is automatically canceled when the FSM exits the current state.
+"""
+function after(action, b::FSMBehavior, delay)
+  waker = WakerBehavior((_, _) -> action(), round(Int, delay*1000))
+  push!(b.wakers, waker)
+  add(b.agent, waker)
+  nothing
+end
+
+"""
+    onenter(a::Agent, b::FSMBehavior, state::FSMState)
+
+Callback method invoked when the FSM enters a state.
+"""
+onenter(a::Agent, b::FSMBehavior, state::FSMState) = nothing
+
+"""
+    action(a::Agent, b::FSMBehavior, state::FSMState)
+
+Callback method invoked cyclically during a state. The default implementation
+calls `block(b)` to block the behavior until an event occurs.
+"""
+action(a::Agent, b::FSMBehavior, state::FSMState) = block(b)
+
+"""
+    onexit(a::Agent, b::FSMBehavior, state::FSMState)
+
+Callback method invoked when the FSM exits a state.
+"""
+onexit(a::Agent, b::FSMBehavior, state::FSMState) = nothing
+
+"""
+    onevent(a::Agent, b::FSMBehavior, state::FSMState, event)
+
+Callback method invoked when an event is triggered in the FSM.
+"""
+onevent(a::Agent, b::FSMBehavior, state, event) = nothing
+
+# FSM behavior main loop
+function action(b::FSMBehavior)
+  try
+    while !done(b)
+      if b.block === nothing
+        try
+          b.state == b.prevstate && _mutex_call(action, b.agent, b, b.state)
```

Comment:
General qn : why do we need `_mutex_call` in `Fjage.jl`?

Reply:
I think the idea is that you shouldn't need mutexes in user code. The framework code will always need some synchronisation primitives to mediate between the agents. But this begs the question whether `action()` shouldn't be considered user code. The synchronisation would then have to be moved up one level to where this `action()` method is called.

Reply:
The framework code needs mutexes because the underlying implementation of behaviors use @async tasks, and they can be preempted at various implicit yield points. The behavior user shouldn't see this and should be able to write atomic user methods, and so we need the mutexes.

The methods we protect are the ones that are meant to be overridden by users.